# 异步

## Promise

### API

- resolve() - 返回一个状态由给定 value 决定的 Promise 对象。
- reject() - 返回一个状态为失败的 Promise 对象。
- all() - 处理 Promise 集合，返回其结果集合，当有一个 Promise 失败，即停止。
- race() - 处理 Promise 集合，当有一个 Promise 处理完即返回。
- any() - 接收一个 Promise 对象的集合，当其中的一个 promise 完成，就返回那个已经有完成的 promise 的值。
- allSettled() - 处理 Promise 集合，返回一个promise，该 promise 在所有 promise 完成后完成。并带有一个对象数组，每个对象对应每个 promise 的结果。

### 实现

按照 [Promise/A+规范](https://promisesaplus.com/) 实现，完成之后可以通过 `promises-aplus-tests` 插件进行测试

```js
// 三种状态
const PENDING = "pending";
const FULFILLED = "fulfilled";
const REJECTED = "rejected";

class MyPromise {
  constructor(excutor) {
    this.status = PENDING; // 当前状态
    this.value = undefined; // 当前值
    this.reason = undefined; // 当前错误原因
    this.onFulfilledCallbacks = []; // 成功回调
    this.onRejectedCallbacks = []; // 失败回调

    const resolve = value => {
      // 传入 Promise 的情况
      if (value instanceof MyPromise) {
        return value.then(resolve, reject);
      }
      // 无法模仿微任务，选择使用宏任务
      setTimeout(() => {
        if (this.status === PENDING) {
          // 成功状态
          this.status = FULFILLED;
          this.value = value;
          this.onFulfilledCallbacks.forEach(cb => cb(this.value));
        }
      });
    };

    const reject = reason => {
      setTimeout(() => {
        if (this.status === PENDING) {
          this.status = REJECTED;
          this.reason = reason;
          this.onRejectedCallbacks.forEach(cb => cb(this.reason));
        }
      });
    };

    try {
      excutor(resolve, reject);
    } catch (e) {
      reject(e);
    }
  }

  then(onFulfilled, onRejected) {
    let newPromise;
    onFulfilled =
      typeof onFulfilled === "function" ? onFulfilled : value => value;
    onRejected =
      typeof onRejected === "function"
        ? onRejected
        : reason => {
            throw reason;
          };
    switch (this.status) {
      // 成功
      case FULFILLED: {
        return (newPromise = new MyPromise((resolve, reject) => {
          // 为了访问 newPromise 变量
          setTimeout(() => {
            try {
              let x = onFulfilled(this.value);
              promiseResolutionProcedure(newPromise, x, resolve, reject); // 新的promise resolve 上一个onFulfilled的返回值
            } catch (e) {
              reject(e); // 捕获前面onFulfilled中抛出的异常 then(onFulfilled, onRejected);
            }
          });
        }));
      }
      // 失败
      case REJECTED: {
        return (newPromise = new MyPromise((resolve, reject) => {
          setTimeout(() => {
            try {
              let x = onRejected(this.reason);
              promiseResolutionProcedure(newPromise, x, resolve, reject);
            } catch (e) {
              reject(e);
            }
          });
        }));
      }
      // 等待
      case PENDING: {
        // 当异步调用resolve/rejected时 将onFulfilled/onRejected收集暂存到集合中
        return (newPromise = new MyPromise((resolve, reject) => {
          this.onFulfilledCallbacks.push(value => {
            try {
              let x = onFulfilled(value);
              promiseResolutionProcedure(newPromise, x, resolve, reject);
            } catch (e) {
              reject(e);
            }
          });
          this.onRejectedCallbacks.push(reason => {
            try {
              let x = onRejected(reason);
              promiseResolutionProcedure(newPromise, x, resolve, reject);
            } catch (e) {
              reject(e);
            }
          });
        }));
      }
    }
  }

  catch(onRejected) {
    return this.then(null, onRejected);
  }

  static all(promises) {
    let count = 0;
    let values = [];
    return new MyPromise((resolve, reject) => {
      promises.forEach((promise, index) => {
        promise.then(value => {
          // 存储 promise 的值
          values[index] = value;
          // 执行完后返回
          if (++count === promises.length) {
            resolve(values);
          }
        }, reject);
      });
    });
  }

  static race(promises) {
    return new MyPromise((resolve, reject) => {
      promises.forEach((promise, index) => {
        promise.then(resolve, reject);
      });
    });
  }

  static resolve(value) {
    return new MyPromise(resolve => {
      resolve(value);
    });
  }

  static reject(reason) {
    return new MyPromise((resolve, reject) => {
      reject(reason);
    });
  }
}

/**
 * 对 resolve 进行改造增强 针对 resolve 中不同值情况 进行处理
 * @param  {promise} promise2 promise1.then 方法返回的新的 promise 对象
 * @param  {[type]} x         promise1 中 onFulfilled 的返回值
 * @param  {[type]} resolve   promise2 的 resolve 方法
 * @param  {[type]} reject    promise2 的 reject 方法
 */
function promiseResolutionProcedure(promise2, x, resolve, reject) {
  // 如果 promise 和 x 引用相同，reject
  if (promise2 === x) {
    return reject(new TypeError("循环引用"));
  }
  // 如果x是一个 promise 对象
  if (x instanceof MyPromise) {
    // 如果 x 是 pending 状态，promise 必须保持 pending 直至 x 完成
    if (x.status === PENDING) {
      x.then(
        y => {
          promiseResolutionProcedure(promise2, y, resolve, reject);
        },
        reason => {
          reject(reason);
        }
      );
    } else {
      // 如果 x 完成，则 promise 得到相同的状态和值
      x.then(resolve, reject);
    }
  } else if (x != null && (typeof x === "object" || typeof x === "function")) {
    // 避免多次调用
    let called = false;
    try {
      let then = x.then;
      // 是否是 thenable 对象（具有 then 方法的对象或函数）
      if (typeof then === "function") {
        then.call(
          x,
          y => {
            // 如果已经调用，忽略
            if (called) return;
            called = true;
            promiseResolutionProcedure(promise2, y, resolve, reject);
          },
          reason => {
            if (called) return;
            called = true;
            reject(reason);
          }
        );
      } else {
        // 如果是一个普通对象或函数，直接 resolve
        resolve(x);
      }
    } catch (e) {
      if (called) return;
      called = true;
      reject(e);
    }
  } else {
    // 如果不是一个对象或函数，直接 resolve
    resolve(x);
  }
}

// npm i promises-aplus-tests
// promises-aplus-tests Promise.js

MyPromise.deferred = function () {
  // 延迟对象
  let defer = {};
  defer.promise = new MyPromise((resolve, reject) => {
    defer.resolve = resolve;
    defer.reject = reject;
  });
  return defer;
};

module.exports = MyPromise;
```

## Generator

Generator 函数是 ES6 提供的一种异步编程解决方案，与传统函数是完全不同的机制。

### 基本用法

```js
function f1() {
  console.log("f1 run");
  return "f1";
}

function f2() {
  console.log("f2 run");
  return "f2";
}

function* helloWorld() {
  console.log("start")
  const hello = yield f1();
  const world = yield f2();
  console.log("end")
  return `${hello} ${world}`;
}

const say = helloWorld();
console.log(say); // Object [Generator] {}

const step1 = say.next(); // start f1 run
console.log(step1); // { value: 'f1', done: false }

const step2 = say.next("Hello"); // f2 run
console.log(step2); // { value: 'f2', done: false }

const step3 = say.next("World"); // end
console.log(step3); // { value: 'Hello World', done: true }
```

1. 通过 `function*` 进行函数声明，通过 `yield` 表达式定义函数内部的状态；
2. 执行 `Generator` 函数可以得到一个 `Generator` 对象，其构造函数的原型 `GeneratorFunction ` 上主要有 `next()` 、`throw()` 、 `return()` 三个方法。
3. 通过调用 `next()` 方法，函数会执行到一个 `yield` 表达式，可通过多次调用将函数执行完毕，返回结果为一个对象 `{ value: any, done: Boolean }` ，`value` 为 `yield` 表达式生成的值，`done` 则表示是否执行完毕，`next()` 可以通过传入一个参数来作为上一个 `yield` 表达式的返回值。

### 结合 Promise 进行异步处理

一个例子，对用户信息进行修改，修改成功后再重新获取用户信息，对比使用 `Promise` 和 `Generator` 处理的优劣。

```js
// step 1
function setUser() {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve("用户信息修改成功");
    }, 0);
  });
}

// step 2
function fetchUser() {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      reject("获取用户信息超时");
    }, 600);
  });
}
```

`Promise` 版

```js
setUser()
  .then((res) => {
    console.log(res);
  })
  .then((res) => {
    return fetchUser();
  })
  .then((res) => {
    console.log(res);
  })
  .catch((e) => console.log(e));
```

通过链式调用依次处理异步方法，可以避免回调地狱的问题，但是如果场景复杂度上来了，代码会显得特别冗余，一屏幕 `then()` 感受一下。

`Generator` 版

```js
function* initData() {
  const status = yield setUser();
  console.log(status);
  const useInfo = yield fetchUser();
  console.log(useInfo);
}
```

代码是用同步的写法去处理异步操作，语义明确，但是需要我们手动执行，手动执行又分手动挡和自动挡两种。

手动挡，手动的方式一步步执行。

```js
// 手动执行 Genorator
const init = initData();

const step1 = init.next().value;
step1
  .then((res) => {
    const step2 = init.next(res).value;
    step2.then((res) => init.next(res)).catch((e) => console.log(e));
  })
  .catch((e) => console.log(e));
```

自动挡，通过递归的方式不断调用 `next()` ，再通过 `done` 判断是否结束递归。

```js
function autoRun(gen) {
  const g = gen();
  function next(data) {
    const result = g.next(data);
    if (result.done) return result.value;
    result.value
      .then(function (data) {
        next(data);
      })
      .catch((e) => console.log(e));
  }
  next();
}

autoRun(initData)
```

通过结合 `Promise` 去完成 `Generator` 函数的执行，将 `Promise` 的链式调用逻辑封装起来，开发者只需要关注异步方法的调用顺序，简化了异步操作，但是函数执行过程的处理难度也不容忽视。

## async/await

ES2017标准引入了 `async/await` 用于异步的处理。

继续使用上面的那个例子，使用 `async/await` 去进行处理：

```js
function setUser() {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve("用户信息修改成功");
    }, 0);
  });
}

function fetchUser() {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      reject("获取用户信息超时");
    }, 600);
  });
}

async function initData() {
  try {
    const res = await setUser();
    console.log(res);
    const info = await fetchUser();
    console.log(info);
  } catch (e) {
    console.log(e);
  }
}

initData()
```

继承了 `Generator` 的优点，同时也避免了执行过程管理的困难。

### 实现原理

我们可以将使用 `async/await` 代码使用 `babel` 进行编译，即可知道其降级实现；通过编译，可以看出 `async/await` 只是 `Generator` 函数的语法糖，但是其内置了执行器，不需要开发者去手动处理函数的执行过程。

```js
function setUser() {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve("用户信息修改成功");
    }, 0);
  });
}

function fetchUser() {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      reject("获取用户信息超时");
    }, 600);
  });
}

// core
function asyncToGenerator(generatorFn) {
  return function () {
    // 缓存 Generator 对象
    const gen = generatorFn.apply(this, arguments);
    return new Promise((resolve, reject) => {
      /**
       * @param {"throw"|"next"} key Generator 方法
       * @param {*} arg 上一个 yield 表达式的值
       */
      function asyncGeneratorStep(key, arg) {
        let result;
        try {
          result = gen[key](arg);
        } catch (e) {
          return reject(e);
        }
        // value 是一个 Promise 对象
        const { value, done } = result;
        // 是否停止递归
        if (done) {
          return resolve(value);
        } else {
          return Promise.resolve(value)
            .then((res) => asyncGeneratorStep("next", res))
            .catch((e) => asyncGeneratorStep("throw", e));
        }
      }
      asyncGeneratorStep("next");
    });
  };
}

function initData() {
  return asyncToGenerator(function* () {
    try {
      const res = yield setUser();
      console.log(res);
      const info = yield fetchUser();
      console.log(info);
    } catch (e) {
      console.log(e);
    }
  }).apply(this, arguments);
}

initData();
```

## 参考

- [Promise详解与实现（Promise/A+规范）](https://www.jianshu.com/p/459a856c476f)
- [Promise/A+规范](https://promisesaplus.com/)


